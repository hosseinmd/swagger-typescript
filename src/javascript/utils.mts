import { Schema, Parameter, Config } from "../types.mjs";
import { getJsdoc } from "../utilities/jsdoc.mjs";
import { isAscending } from "../utils.mjs";

function getPathParams(parameters?: Parameter[]): Parameter[] {
  return (
    parameters?.filter(({ in: In }) => {
      return In === "path";
    }) || []
  );
}

function getHeaderParams(parameters: Parameter[] | undefined, config: Config) {
  const queryParamsArray =
    parameters?.filter(({ in: In, name }) => {
      return In === "header" && !config.ignore?.headerParams?.includes(name);
    }) || [];

  const params = getObjectType(queryParamsArray, config);

  return {
    params,
    isNullable: queryParamsArray.every(({ schema = {} }) => !schema.required),
  };
}

function toPascalCase(str: string): string {
  return `${str.substring(0, 1).toUpperCase()}${str.substring(1)}`;
}
function replaceWithUpper(str: string, sp: string) {
  let pointArray = str.split(sp);
  pointArray = pointArray.map((point) => toPascalCase(point));

  return pointArray.join("");
}

function generateServiceName(
  endPoint: string,
  method: string,
  operationId: string | undefined,
  config: Config,
): string {
  const { methodName, prefix = "" } = config;

  const _endPoint = endPoint.replace(new RegExp(`^${prefix}`, "i"), "");
  const path = getSchemaName(_endPoint);

  const methodNameTemplate = getTemplate(methodName, operationId);

  const serviceName = template(methodNameTemplate, {
    path,
    method,
    ...(operationId ? { operationId } : {}),
  });
  return serviceName;
}

function getTemplate(methodName?: string, operationId?: string) {
  const defaultTemplate = "{method}{path}";
  if (!methodName) {
    return defaultTemplate;
  }

  const hasMethodNameOperationId = /(\{operationId\})/i.test(methodName);

  if (hasMethodNameOperationId) {
    return operationId ? methodName : defaultTemplate;
  }

  return methodName;
}

const TYPES = {
  integer: "number",
  number: "number",
  boolean: "boolean",
  object: "object",
  string: "string",
  array: "array",
};

function getDefineParam(
  name: string,
  required: boolean = false,
  schema: Schema | undefined,
  config: Config,
  description?: string,
): string {
  return getParamString(name, required, getTsType(schema, config), description);
}

function getParamString(
  name: string,
  required: boolean = false,
  type: string,
  description?: string,
  isPartial?: boolean,
): string {
  return `${getJsdoc({
    description,
  })}${name}${required ? "" : "?"}: ${isPartial ? `Partial<${type}>` : type}`;
}
//x-nullable
function normalizeObjectPropertyNullable(
  propertyName: string,
  schema: Schema,
  required?: string[],
) {
  if (schema.nullable !== undefined) {
    return schema.nullable;
  }
  if (schema["x-nullable"] !== undefined) {
    return schema["x-nullable"];
  }
  if (required) {
    return !required.includes(propertyName);
  }
  return true;
}

function getTsType(
  schema: undefined | true | {} | Schema,
  config: Config,
): string {
  if (isTypeAny(schema)) {
    return "any";
  }

  const {
    type,
    $ref,
    enum: Enum,
    items,
    properties,
    oneOf,
    additionalProperties,
    required,
    allOf,
  } = schema as Schema;

  if ($ref) {
    const refArray = $ref.split("/");
    if (refArray[refArray.length - 2] === "requestBodies") {
      return `RequestBody${getRefName($ref)}`;
    } else {
      return getRefName($ref);
    }
  }
  if (Enum) {
    return `${Enum.map((e) => JSON.stringify(e)).join(" | ")}`;
  }

  if (items) {
    return `${getTsType(items, config)}[]`;
  }

  let result = "";

  if (properties) {
    result += getObjectType(
      Object.entries(properties).map(([pName, _schema]) => ({
        schema: {
          ..._schema,
          nullable: normalizeObjectPropertyNullable(pName, _schema, required),
        },
        name: pName,
      })),
      config,
    );
  }

  if (oneOf) {
    result = `${result} & (${oneOf
      .map((t) => `(${getTsType(t, config)})`)
      .join(" | ")})`;
  }

  if (allOf) {
    result = `${result} & (${allOf
      .map((_schema) => getTsType(_schema, config))
      .join(" & ")})`;
  }

  if (type === "object" && !result) {
    if (additionalProperties) {
      return `{[x: string]: ${getTsType(additionalProperties, config)}}`;
    }

    return "{[x in string | number ]: any}";
  }

  return result || TYPES[type as keyof typeof TYPES];
}

function getObjectType(
  parameter: { schema?: Schema; name: string }[],
  config: Config,
) {
  const object = parameter
    .sort(
      (
        { name, schema: { nullable } = {} },
        { name: _name, schema: { nullable: _nullable } = {} },
      ) => {
        if (!nullable && _nullable) {
          return -1;
        } else if (nullable && !_nullable) {
          return 1;
        }

        return isAscending(name, _name);
      },
    )
    .reduce(
      (
        prev,
        {
          schema: {
            deprecated,
            "x-deprecatedMessage": deprecatedMessage,
            example,
            nullable,
          } = {},
          schema,
          name,
        },
      ) => {
        return `${prev}${getJsdoc({
          ...schema,
          deprecated:
            deprecated || deprecatedMessage ? deprecatedMessage : undefined,
          example,
        })}"${name}"${nullable ? "?" : ""}: ${getTsType(schema, config)};`;
      },
      "",
    );

  return object ? `{${object}}` : "";
}
function getSchemaName(name: string): string {
  ["/", ".", "`", "[", "]", "-", "*", "{", "}"].forEach((str) => {
    name = replaceWithUpper(name, str);
  });

  return name;
}

function getRefName($ref: string): string {
  const parts = $ref.split("/").pop();
  return getSchemaName(parts || "");
}

function getParametersInfo(
  parameters: Parameter[] | undefined,
  type: "query" | "header",
) {
  const params =
    parameters?.filter(({ in: In }) => {
      return In === type;
    }) || [];

  return {
    params,
    exist: params.length > 0,
    isNullable: !params.some(
      ({ schema, required }) =>
        //swagger 2
        required ||
        // openapi 3
        schema?.required,
    ),
  };
}

function isTypeAny(type: true | undefined | {} | Schema) {
  if (type === true) {
    return true;
  }

  if (typeof type === "object" && Object.keys(type).length <= 0) {
    return true;
  }

  if (!type || (type as Schema).AnyValue) {
    return true;
  }

  return false;
}

/** Used to replace {name} in string with obj.name */
function template(str: string, obj: { [x: string]: string } = {}) {
  Object.entries(obj).forEach(([key, value]) => {
    const re = new RegExp(`{${key}}`, "i");
    str = str.replace(re, value);
  });

  const re = new RegExp("{*}", "g");
  if (re.test(str)) {
    throw new Error(`methodName: Some A key is missed "${str}"`);
  }
  return str;
}

export {
  getPathParams,
  getHeaderParams,
  generateServiceName,
  getTsType,
  getRefName,
  isAscending,
  getDefineParam,
  getParamString,
  getParametersInfo,
  isTypeAny,
  template,
  toPascalCase,
  getSchemaName,
};
