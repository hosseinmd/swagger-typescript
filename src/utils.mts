import { Config } from "./types.mjs";
import { exec } from "child_process";

function isAscending(a: string, b: string) {
  if (a > b) {
    return 1;
  }
  if (b > a) {
    return -1;
  }
  return 0;
}

function majorVersionsCheck(expectedV: string, inputV?: string) {
  if (!inputV) {
    throw new Error(
      `Swagger-Typescript working with openApi v3/ swagger v2, seem your json is not openApi openApi v3/ swagger v2`,
    );
  }

  const expectedVMajor = expectedV.split(".")[0];
  const inputVMajor = inputV.split(".")[0];
  function isValidPart(x: string) {
    return /^\d+$/.test(x);
  }
  if (!isValidPart(expectedVMajor) || !isValidPart(inputVMajor)) {
    throw new Error(
      `Swagger-Typescript working with openApi v3/ swagger v2 your json openApi version is not valid "${inputV}"`,
    );
  }

  const expectedMajorNumber = Number(expectedVMajor);
  const inputMajorNumber = Number(inputVMajor);

  if (expectedMajorNumber <= inputMajorNumber) {
    return;
  }

  throw new Error(
    `Swagger-Typescript working with openApi v3/ swagger v2 your json openApi version is ${inputV}`,
  );
}

function isMatchWholeWord(stringToSearch: string, word: string) {
  return new RegExp("\\b" + word + "\\b").test(stringToSearch);
}

async function getCurrentUrl({ url, branch: branchName }: Config) {
  const urls = url as Exclude<Config["url"], undefined | string>;
  if (!branchName) {
    branchName = await execAsync("git branch --show-current");

    branchName = branchName?.split("/")[0];

    branchName = urls.find((item) => branchName === item.branch)?.branch;
  }
  if (!branchName) {
    branchName = (await getSourceBranch()).find((treeItem) =>
      urls.find((item) => treeItem === item.branch),
    ) as string;
  }

  const currentUrl =
    urls.find((item) => branchName === item.branch)?.url || urls[0].url;

  return currentUrl;
}

async function getSourceBranch() {
  const result = await execAsync('git log --format="%D"');
  const branchesTree = result
    .split("\n")
    .flatMap((item) => item.split(", "))
    .map((branch) => {
      branch = branch.trim();

      branch = branch.replace("HEAD -> ", "");
      branch = branch.trim();

      return branch;
    });

  return branchesTree;
}

async function execAsync(command: string) {
  return new Promise<string>((resolve, reject) => {
    const child = exec(command, (error, stdout) => {
      child.kill();
      if (error) {
        reject(error);
        return;
      }

      resolve(stdout);
    });
  });
}

export { getCurrentUrl, majorVersionsCheck, isAscending, isMatchWholeWord };
