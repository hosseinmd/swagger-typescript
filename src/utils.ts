import { Schema, Parameter, SwaggerConfig, JsdocAST } from "./types";

function getPathParams(parameters?: Parameter[]): Parameter[] {
  return (
    parameters?.filter(({ in: In }) => {
      return In === "path";
    }) || []
  );
}

function getHeaderParams(parameters?: Parameter[], config?: SwaggerConfig) {
  return getParams(parameters, "header", config?.ignore?.headerParams);
}

function getParams(
  parameters: Parameter[] | undefined,
  type: "query" | "header",
  ignoreParams?: string[],
) {
  const queryParamsArray =
    parameters?.filter(({ in: In, name }) => {
      return In === type && !ignoreParams?.includes(name);
    }) || [];

  const params = getObjectType(queryParamsArray);

  return {
    params,
    hasNullable: queryParamsArray.every(({ schema }) => schema.nullable),
  };
}

function toPascalCase(str: string): string {
  return `${str.substring(0, 1).toUpperCase()}${str.substring(1)}`;
}

function generateServiceName(
  endPoint: string,
  method: string,
  operationId: string | undefined,
  config: SwaggerConfig,
): string {
  function replaceWithUpper(str: string, sp: string) {
    let pointArray = str.split(sp);
    pointArray = pointArray.map((point) => toPascalCase(point));

    return pointArray.join("");
  }

  const path = replaceWithUpper(
    replaceWithUpper(
      replaceWithUpper(replaceWithUpper(endPoint, "/"), "{"),
      "}",
    ),
    "-",
  );

  const { methodName } = config;
  const hasMethodNameOperationId = /(\{operationId\})/i.test(methodName);
  let methodNameTemplate = hasMethodNameOperationId
    ? operationId
      ? methodName
      : false
    : methodName;
  methodNameTemplate = methodNameTemplate || "{method}{path}";

  const serviceName = template(methodNameTemplate, {
    path,
    method,
    ...(operationId ? { operationId } : {}),
  });
  return serviceName;
}

const TYPES = {
  integer: "number",
  number: "number",
  boolean: "boolean",
  object: "object",
  string: "string",
  array: "array",
};

function getDefineParam(
  name: string,
  required: boolean = false,
  schema: Schema,
  description?: string,
): string {
  return getParamString(name, required, getTsType(schema), description);
}
function getParamString(
  name: string,
  required: boolean = false,
  type: string,
  description?: string,
): string {
  return `${getJsdoc({
    description,
  })}${name}${required ? "" : "?"}: ${type}`;
}

function getTsType(schema: true | {} | Schema): string {
  if (isTypeAny(schema)) {
    return "any";
  }

  const {
    type,
    $ref,
    enum: Enum,
    items,
    properties,
    oneOf,
    additionalProperties,
    required,
  } = schema as Schema;
  let tsType = TYPES[type as keyof typeof TYPES];

  if (type === "object" && additionalProperties) {
    tsType = `{[x: string]: ${getTsType(additionalProperties)}}`;
  }
  if ($ref) {
    const refArray = $ref.split("/");
    if (refArray[refArray.length - 2] === "requestBodies") {
      tsType = `RequestBody${getRefName($ref)}`;
    } else {
      tsType = getRefName($ref);
    }
  }
  if (Enum) {
    tsType = `${Enum.map((t) => `"${t}"`).join(" | ")}`;
  }

  if (items) {
    tsType = `${getTsType(items)}[]`;
  }

  if (oneOf) {
    tsType = `${oneOf.map((t) => `(${getTsType(t)})`).join(" | ")}`;
  }

  if (properties) {
    tsType = getObjectType(
      Object.entries(properties).map(([pName, _schema]) => ({
        schema: {
          ..._schema,
          nullable: !required?.find((name) => name === pName),
        },
        name: pName,
      })),
    );
  }

  return tsType;
}

function getObjectType(parameter: { schema: Schema; name: string }[]) {
  const object = parameter
    .sort(
      (
        { name, schema: { nullable } },
        { name: _name, schema: { nullable: _nullable } },
      ) => {
        if (!nullable && _nullable) {
          return -1;
        } else if (nullable && !_nullable) {
          return 1;
        }

        return isAscending(name, _name);
      },
    )
    .reduce(
      (
        prev,
        {
          schema: {
            title,
            description,
            deprecated,
            "x-deprecatedMessage": deprecatedMessage,
            example,
            nullable,
          },
          schema,
          name,
        },
      ) => {
        return `${prev}${getJsdoc({
          description: assignToDescription({
            description,
            title,
            format: schema.format,
            maxLength: schema.maxLength,
            min: schema.min,
            max: schema.max,
            pattern: schema.pattern,
          }),
          tags: {
            deprecated: {
              value: Boolean(deprecated),
              description: deprecatedMessage,
            },
            example,
          },
        })}${name}${nullable ? "?" : ""}: ${getTsType(schema)};`;
      },
      "",
    );

  return object ? `{${object}}` : "";
}

function getRefName($ref: string): string {
  return $ref.replace(/(#\/components\/\w+\/)/g, "");
}

function isAscending(a: string, b: string) {
  if (a > b) {
    return 1;
  }
  if (b > a) {
    return -1;
  }
  return 0;
}

function getParametersInfo(
  parameters: Parameter[] | undefined,
  type: "query" | "header",
) {
  const params =
    parameters?.filter(({ in: In }) => {
      return In === type;
    }) || [];

  return {
    params,
    exist: params.length > 0,
    isNullable: params.every(({ schema }) => schema.nullable),
  };
}

function assignToDescription({
  description,
  title,
  format,
  maxLength,
  max,
  min,
  pattern,
}: {
  title?: string;
  description?: string;
  format?: string;
  pattern?: string;
  maxLength?: number;
  min?: number;
  max?: number;
}) {
  return `${
    title
      ? `
  * ${title}
  *`
      : ""
  }${
    description
      ? `
  * ${description}
  `
      : ""
  }${
    format
      ? `
  *    Format: ${format}`
      : ""
  }${
    maxLength
      ? `
  *    maxLength: ${maxLength}`
      : ""
  }${
    min
      ? `
  *    min: ${min}`
      : ""
  }${
    max
      ? `
  *    max: ${max}`
      : ""
  }${
    pattern
      ? `
  *    pattern: ${pattern}`
      : ""
  }`;
}

function getJsdoc({
  description,
  tags: { deprecated, example } = {},
}: JsdocAST) {
  return deprecated?.value || description || example
    ? `
      /**${
        description
          ? `
      * ${description}`
          : ""
      }${
        deprecated?.value
          ? `
      * @deprecated ${deprecated.description || ""}`
          : ""
      }${
        example
          ? `
      * @example 
      *   ${example}`
          : ""
      }
      */
`
    : "";
}

function majorVersionsCheck(expectedV: string, inputV?: string) {
  if (!inputV) {
    throw new Error(
      `Swagger-Typescript working with openApi v3, seem your json is not openApi v3`,
    );
  }

  const expectedVMajor = expectedV.split(".")[0];
  const inputVMajor = inputV.split(".")[0];
  function isValidPart(x: string) {
    return /^\d+$/.test(x);
  }
  if (!isValidPart(expectedVMajor) || !isValidPart(inputVMajor)) {
    throw new Error(
      `Swagger-Typescript working with openApi v3 your json openApi version is not valid "${inputV}"`,
    );
  }

  const expectedMajorNumber = Number(expectedVMajor);
  const inputMajorNumber = Number(inputVMajor);

  if (expectedMajorNumber === inputMajorNumber) {
    return;
  } else if (expectedMajorNumber < inputMajorNumber) {
    return;
  }

  throw new Error(
    `Swagger-Typescript working with openApi v3 your json openApi version is ${inputV}`,
  );
}

function isTypeAny(type: true | {} | Schema) {
  if (type === true) {
    return true;
  }

  if (typeof type === "object" && Object.keys(type).length <= 0) {
    return true;
  }

  if ((type as Schema).AnyValue) {
    return true;
  }

  return false;
}

/** Used to replace {name} in string with obj.name */
function template(str: string, obj: { [x: string]: string } = {}) {
  Object.entries(obj).forEach(([key, value]) => {
    const re = new RegExp(`{${key}}`, "i");
    str = str.replace(re, value);
  });

  const re = new RegExp("{*}", "g");
  if (re.test(str)) {
    throw new Error(`methodName: Some A key is missed "${str}"`);
  }
  return str;
}

export {
  majorVersionsCheck,
  getPathParams,
  getHeaderParams,
  generateServiceName,
  getTsType,
  getRefName,
  isAscending,
  getDefineParam,
  getParamString,
  getParametersInfo,
  getJsdoc,
  isTypeAny,
  template,
  toPascalCase,
};
