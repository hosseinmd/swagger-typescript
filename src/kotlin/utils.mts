import { Schema, Parameter, Config } from "../types.mjs";
import { getJsdoc } from "../utilities/jsdoc.mjs";
import { isAscending } from "../utils.mjs";

function getPathParams(parameters?: Parameter[]): Parameter[] {
  return (
    parameters?.filter(({ in: In }) => {
      return In === "path";
    }) || []
  );
}

function getHeaderParams(parameters: Parameter[] | undefined, config: Config) {
  const params =
    parameters?.filter(({ in: In, name }) => {
      return In === "header" && !config.ignore?.headerParams?.includes(name);
    }) || [];

  return {
    params,
    isNullable: params.every(({ schema = {} }) => !schema.required),
  };
}

function toPascalCase(str: string): string {
  return `${str.substring(0, 1).toUpperCase()}${str.substring(1)}`;
}
function replaceWithUpper(str: string, sp: string) {
  let pointArray = str.split(sp);
  pointArray = pointArray.map((point) => toPascalCase(point));

  return pointArray.join("");
}

function generateServiceName(
  endPoint: string,
  method: string,
  operationId: string | undefined,
  config: Config,
): string {
  const { methodName, prefix = "" } = config;

  const _endPoint = endPoint.replace(new RegExp(`^${prefix}`, "i"), "");
  const path = getSchemaName(_endPoint);

  const methodNameTemplate = getTemplate(methodName, operationId);

  const serviceName = template(methodNameTemplate, {
    path,
    method,
    ...(operationId ? { operationId } : {}),
  });
  return serviceName;
}

function getTemplate(methodName?: string, operationId?: string) {
  const defaultTemplate = "{method}{path}";
  if (!methodName) {
    return defaultTemplate;
  }

  const hasMethodNameOperationId = /(\{operationId\})/i.test(methodName);

  if (hasMethodNameOperationId) {
    return operationId ? methodName : defaultTemplate;
  }

  return methodName;
}

const TYPES = {
  integer: "Int",
  number: "Long",
  boolean: "Boolean",
  object: "Any",
  string: "String",
  array: "List<Any>",
};

function getDefineParam(
  name: string,
  required: boolean = false,
  schema: Schema | undefined,
  config: Config,
  description?: string,
): string {
  return getParamString(
    name,
    required,
    getKotlinType(schema, config),
    description,
  );
}

function getDefinitionBody(
  name: string,
  schema: Schema | undefined,
  config: Config,
  description?: string,
): string {
  const type = getKotlinType(schema, config);
  return `${getJsdoc({
    description,
  })}@Body ${name}: ${type}`;
}

function getHeaderString(
  name: string,
  required: boolean = false,
  type: string,
  description?: string,
): string {
  return `${
    //   getJsdoc({
    //   description,
    // })
    ""
  }
@Headers("Content-Type: ${type}")`;
}

function getParamString(
  name: string,
  required: boolean = false,
  type: string,
  description?: string,
  isPartial?: boolean,
): string {
  return `${
    //   getJsdoc({
    //   description,
    // })
    ""
  }@Path("${name}") ${name}: ${type}${required ? "" : "?"}`;
}
//x-nullable
function normalizeObjectPropertyNullable(
  propertyName: string,
  schema: Schema,
  required?: string[],
) {
  if (schema.nullable !== undefined) {
    return schema.nullable;
  }
  if (schema["x-nullable"] !== undefined) {
    return schema["x-nullable"];
  }
  if (required) {
    return !required.includes(propertyName);
  }
  return true;
}

function getClassBody(
  schema: undefined | true | {} | Schema,
  config: Config,
): string {
  if (isTypeAny(schema)) {
    return "Any";
  }

  const { properties, required } = schema as Schema;

  if (properties) {
    return getObjectType(
      Object.entries(properties).map(([pName, _schema]) => ({
        schema: {
          ..._schema,
          nullable: normalizeObjectPropertyNullable(pName, _schema, required),
        },
        name: pName,
      })),
      config,
    );
  }

  return "Any";
}

function getKotlinType(
  schema: undefined | true | {} | Schema,
  config: Config,
): string {
  if (isTypeAny(schema)) {
    return "Any";
  }

  const {
    type,
    $ref,
    enum: Enum,
    items,
    properties,
    oneOf,
    additionalProperties,
    required,
    allOf,
  } = schema as Schema;

  if ($ref) {
    const refArray = $ref.split("/");
    if (refArray[refArray.length - 2] === "requestBodies") {
      return `RequestBody${getRefName($ref)}`;
    } else {
      return getRefName($ref);
    }
  }
  if (Enum) {
    return "String";
    // return `${Enum.map((t) => `"${t}"`).join(" | ")}`;
  }

  if (items) {
    return `List<${getKotlinType(items, config)}>`;
  }

  let result = "";

  if (properties) {
    result = "Any";
  }

  if (oneOf) {
    result = "Any";
    // result = `${result} & (${oneOf
    //   .map((t) => `(${getTsType(t, config)})`)
    //   .join(" | ")})`;
  }

  if (allOf) {
    result = "Any";

    // result = `${result} & (${allOf
    //   .map((_schema) => getTsType(_schema, config))
    //   .join(" & ")})`;
  }

  if (type === "object" && !result) {
    // if (additionalProperties) {
    //   return `{[x: string]: ${getTsType(additionalProperties, config)}}`;
    // }

    return "Any";
  }

  return result || TYPES[type as keyof typeof TYPES];
}

function getObjectType(
  parameter: { schema?: Schema; name: string }[],
  config: Config,
) {
  const object = parameter
    .sort(
      (
        { name, schema: { nullable } = {} },
        { name: _name, schema: { nullable: _nullable } = {} },
      ) => {
        if (!nullable && _nullable) {
          return -1;
        } else if (nullable && !_nullable) {
          return 1;
        }

        return isAscending(name, _name);
      },
    )
    .reduce(
      (
        prev,
        {
          schema: {
            deprecated,
            "x-deprecatedMessage": deprecatedMessage,
            example,
            nullable,
          } = {},
          schema,
          name,
        },
      ) => {
        const jsdoc = getJsdoc({
          ...schema,
          deprecated:
            deprecated || deprecatedMessage ? deprecatedMessage : undefined,
          example,
        });
        return `${
          prev
            ? `
  ${prev}`
            : ""
        }${jsdoc}
  val ${name}: ${getKotlinType(schema, config)}${nullable ? "?" : ""},`;
      },
      "",
    );

  return object;
}
function getSchemaName(name: string): string {
  ["/", ".", "`", "[", "]", "-", "*", "{", "}"].forEach((str) => {
    name = replaceWithUpper(name, str);
  });

  return name;
}

function getRefName($ref: string): string {
  const parts = $ref.split("/").pop();
  return getSchemaName(parts || "");
}

function getParametersInfo(
  parameters: Parameter[] | undefined,
  type: "query" | "header",
) {
  const params =
    parameters?.filter(({ in: In }) => {
      return In === type;
    }) || [];

  return {
    params,
    exist: params.length > 0,
    isNullable: !params.some(
      ({ schema, required }) =>
        //swagger 2
        required ||
        // openapi 3
        schema?.required,
    ),
  };
}

function isTypeAny(type: true | undefined | {} | Schema) {
  if (type === true) {
    return true;
  }

  if (typeof type === "object" && Object.keys(type).length <= 0) {
    return true;
  }

  if (!type || (type as Schema).AnyValue) {
    return true;
  }

  return false;
}

/** Used to replace {name} in string with obj.name */
function template(str: string, obj: { [x: string]: string } = {}) {
  Object.entries(obj).forEach(([key, value]) => {
    const re = new RegExp(`{${key}}`, "i");
    str = str.replace(re, value);
  });

  const re = new RegExp("{*}", "g");
  if (re.test(str)) {
    throw new Error(`methodName: Some A key is missed "${str}"`);
  }
  return str;
}

export {
  getPathParams,
  getHeaderParams,
  generateServiceName,
  getKotlinType,
  getClassBody,
  getRefName,
  isAscending,
  getDefineParam,
  getParamString,
  getParametersInfo,
  isTypeAny,
  template,
  toPascalCase,
  getSchemaName,
  getDefinitionBody,
  getHeaderString,
};
