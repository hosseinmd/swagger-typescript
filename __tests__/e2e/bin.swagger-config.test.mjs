import { spawn } from "child_process";
import { promises as fs } from "fs";
import path from "path";
import { fileURLToPath } from "url";
import { cleanOutputDir, generate } from "./utils.mjs";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const localConfigPath = path.join(__dirname, "swagger.config.json");

describe("swag-ts CLI E2E Tests", () => {
  //   beforeEach(async () => {
  //     await cleanOutputDir(localConfigPath);
  //   });

  //   afterEach(async () => {
  //     await cleanOutputDir(localConfigPath);
  //   });

  describe("Local OpenAPI Tests", () => {
    test("E2E-LOCAL-001: Django Allauth API Generation", async () => {
      const generatedFiles = await generate(localConfigPath);

      // Create snapshot of all generated files for local OpenAPI

      // Verify critical content exists

      delete generatedFiles["swagger.json"];
      expect(generatedFiles).toMatchSnapshot("local-openapi-generated-files");
    }, 30000);
  });
});
