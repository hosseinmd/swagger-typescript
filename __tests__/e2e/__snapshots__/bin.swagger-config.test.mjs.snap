// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swag-ts CLI E2E Tests Local OpenAPI Tests E2E-LOCAL-001: Django Allauth API Generation: local-openapi-generated-files 1`] = `
Object {
  "config.ts": "",
  "hooks.ts": "",
  "hooksConfig.ts": "",
  "httpRequest.ts": "",
  "services.ts": "",
  "types.ts": "",
}
`;
